### ISSUE TITLE
Thread Time Activities nested when they should not be

### ISSUE THREAD

**Issue by vancem** [Microsoft/perfview#725]
See on vancem1  the WrongCallTree.m.etl trace.  

In the ThreadTime with start stop activities view you will see the 

Activity Request(//1/292/,)

Which has a series of nested activites (293, 298, 299, ...)   These are not nested, they are independent.

**Comment by vancem** [comment]
More details.   In the trace 


Event Name | Time MSec
-- | --
Microsoft-ApplicationInsights-Data/Request/Start | 78,847.776
Microsoft-ApplicationInsights-Data/Request/Stop | 79,126.967

--

**Comment by vancem** [comment]
Note that most of the Request/Start and Request/Stop events are paired as you would expect (but that also seems to be exactly when there is no multiple Async I/O is going on on that thread.  

I have tried to create a repro so I can investigate why the runtime was not emitting the needed events, but had trouble.   Saar is helping me.  

My strategy is to use an async-local to generate events every time Task execution contexts change.   That should highlight the path when this failure happens, which should tell us where the missing event is.

**Comment by vancem** [comment]
Repro instructions:

Create a .NET Core MVC app.   Modify the HomeController.Index() method to be
```
        public async Task<ActionResult> Index()
        {
            int delay = new Random().Next(200) + 200;
            await Task.Delay(delay);
            return View();
        }
```
IMPORTANT: also comment out the line
```
           app.UseHttpsRedirection();
```
In the Startup.Configure method.  Otherwise it tries to use HTTPS instead of just HTTP, and that needlessly complicates the scenario. 

Run the app, and generate load with it using the LoaderGenerator app  with the following params
```
dotnet LoadGenerator.dll http://localhost:5000 200 20
```
Which starts 200 threads each of which produce 20 requests (thus 4000 total request),  The loadGenerator program is at

\\clrmain\public\PerfInvestigations\17-01-25.AspNet.TechEmpower\LoadGenerator

I have a TaskDebugger class that uses asyncLocals to log every time a context changes.  This can be useful (but not needed to repro).  

I would also turn on stacks for all TPL events (normally we turn most of them off), with the perfView qualifer
```
/providers=*Microsoft-DotNet-TaskDebugger:@StacksEnabled=true,.NETTasks:@StacksEnabled=true
```
Finally turning on /dotnetCalls or /dotnetAlloc will give you stacks at many points which can be useful just to understand the flow (but of course it slows things down alot.  Newer version of PerfVIew have /dotnetCallsSampled which may also be use.

In the 'Events view you will find
```
Microsoft-Diagnostics-DiagnosticSource/Activity1Start/Start
Microsoft-Diagnostics-DiagnosticSource/Activity1Start/Stop
```
which show the start-stop events that show up in the ThreadTime with start-stop tasks.  If you look at any particular thread, when you see multiple starts in a row, that is a sign that you are creating the right conditions to find the bug.   If that thread does not have something that indicates that a execution context switched between the two starts, that is a problem (since PerfView will assume they are nested).

On Vancem1 there is a file WrongCallTreeWithEvents.etl.zip which is one such collection.  Note in this trace the TaskDebugger events did not get parsed.  They show up asProvider(5060e28e-307e-5eca-c182-f1446fa6d35e)/EventID(2), which is the 'ContextChanged' event.  

--

**Comment by Xhanti** [comment]
Hi @vancem ,

I have a trace exhibiting incorrectly nested activities from an asp.net framework app using appinsights.
Looking at this thread I just wanted to clarify if this issue is related to using app insights and if so is the fix going to be to perfview or to appinsights? 

Regards

**Comment by vancem** [comment]
This issue above has been fixed with PR https://github.com/Microsoft/perfview/pull/728 (actually only the change on ActivityComputer.cs is needed for this particular issue).  

Basically the issue is that ASP.NET was using 'raw' ThreadPool APIs, and these lower level APIs did not emit some of the thread pool logging that the higher level (Task) APIs did.    There were however events that WERE available, but they were not 'connected up' completely (that was an oversight), and that fix did the final wiring up needed.  

If you build PerfVIew right now it has the fix.  I wanted to wait on some other fixes to get in before publishing the next PerfView.exe on Github, but will bail on that and simply push what is currently checked in in that next day or two if I can't get the other fixes in.

**Comment by vancem** [comment]
The official 2.0.23 version of Perfview available from https://github.com/Microsoft/perfview/blob/master/documentation/Downloading.md has this fix.

**Comment by Xhanti** [comment]
Hi @vancem 

I downloaded the latest release and the issue I was experiencing is still happening.
Perhaps I'm misunderstanding what the **_'Thread Time (With StartStop Activities) (CPU ONLY) Stacks'_** view is supposed to show.

Below is what I see from a trace on one of our production nodes:
![image](https://user-images.githubusercontent.com/2096304/45208830-d9424f80-b28b-11e8-9145-b01254e4f9e9.png)
I would expect each _IISRequest_ to be one activity, then the _AspNetReq_ to be a nested activity and then all other activities to be nested within that accordingly. From the trace it looks like one _IISRequest_ spawns many _IISRequests_ that then begin to nest, 

Is this expected behavior? If so, under what circumstances might this occur?
Could you point me in the direction of where to start looking for what might cause this, I'm interested in understanding this a bit more.
If you need additional information please let me know.

I'm happy to open another issue if you think this is unrelated.

Regards

**Comment by Xhanti** [comment]
Looking at the open issues I found this as well [708](https://github.com/Microsoft/perfview/issues/708).
We are using appInsights as well
